qutip/solver/dysolve_propagator.py

```diff
@@ -0,0 +1,473 @@
+    """
+    A generator of propagator using Dysolve.
+    https://arxiv.org/abs/2012.09282
+
+    Parameters
+    ----------
+    H_0 : Qobj
+        The base hamiltonian of the system.
+
+    X : Qobj
+        A cosine perturbation applied on the system.
+
+    omega : Number
+        The frequency of the cosine perturbation.
+
+    options : dict, optional
+        Extra parameters.
+
+        - "max_order"
+
+            A given integer to indicate the highest order of
+            approximation used to compute the propagators (default is 4).
+            This corresponds to n in eq. (4) of Ref.
+
+        - "a_tol"
+
+            The absolute tolerance used when computing the propagators
+            (default is 1e-10).
+
+    Notes
+    -----
+    The system's hamiltonian must be of the form
+    H = H_0 + cos(omega*t)X for Dysolve to work.
+
+    For the moment, only a cosine perturbation is allowed. Dysolve can
+    manage more exotic perturbations, but this is not implemented yet.
+
+    Experimental.
+
+    """
+
+    def __init__(
+        self,
+        H_0: Qobj,
+        X: Qobj,
+        omega: Number,
+        options: dict[str] = None,
+    ):
+        # System
+        self._eigenenergies, self._basis = H_0.eigenstates()
+        self.H_0 = H_0
+        self.X = X
+        self.omega = omega
+
+        # Times
+        self.t_i = None
+        self.t_f = None
+        self._dt = None
+        self._max_dt = 0.1
+
+        # Options
+        if options is None:
+            self.max_order = 4
+            self.a_tol = 1e-10
+        else:
+            self.max_order = options.get('max_order', 4)
+            self.a_tol = options.get('a_tol', 1e-10)
+
+        self._Sns = None
+        self._dt_Sns = {}
+        self.U = None
+
+    def __call__(self, t_f: Number, t_i: Number = 0) -> Qobj:
+        """
+        Computes the propagator from t_i to t_f. If t_i is not provided,
+        computes the propagator from 0 to t_f.
+
+        Parameters
+        ----------
+        t_f : Number
+            Final time of the evolution.
+
+        t_i : Number, default = 0
+            Initial time of the evolution.
+
+        Returns
+        -------
+        U : Qobj
+            The propagator U(t_f, t_i) from t_i to t_f.
+
+        Notes
+        -----
+        If t_f - t_i > 0.1, splits the evolution into smaller ones
+        to then reconstruct U(t_f, t_i).
+
+        Memoization is used. First call may be slow but the next calls
+        should be faster.
+
+        """
+        self.t_i = t_i
+        self.t_f = t_f
+        time_diff = t_f - t_i
+        n_steps = abs(int(time_diff / self._max_dt))
+
+        U = np.eye(len(self._eigenenergies), dtype=np.complex128)
+
+        self._dt = self._max_dt * np.sign(time_diff)
+        if self._dt not in self._dt_Sns:
+            self._Sns = self._compute_Sns()
+            self._dt_Sns[self._dt] = self._Sns
+        else:
+            self._Sns = self._dt_Sns[self._dt]
```

Comment:
This memorization should be in `_compute_Sns`:
`_Sns = self._compute_Sns(dt)`
This would remove the copy of these lines in the `U_extra` section.
Also, `self._dt` and `self._Sns` could be removed.

`__call__` is the only user facing function, so any attribute that is overwritten should probably not be an attribute but just a variable.